hardware-handler/nightmare_hardware_handler/src/hardware-handler.py

```diff
@@ -13,48 +15,83 @@
 from std_msgs.msg import Header
 
 # robot servo serial ports
-tty_list = ["/dev/ttyUSB0", "/dev/ttyUSB1", "/dev/ttyUSB2", "/dev/ttyUSB3"]
+TTY_LIST = ["/dev/ttyUSB0", "/dev/ttyUSB1", "/dev/ttyUSB2", "/dev/ttyUSB3"]
+NUMBER_OF_LEGS = 6
+NUMBER_OF_SERVOS = 19
+NUMBER_OF_SERVOS_PER_LEG = 3
 
 
 # simple range mapping function
 def fmap(x: float, in_min, in_max, out_min, out_max):
     return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min
 
 
-# servo class to hold servo parameters
+@dataclass
 class Servo:
-    def __init__(self):
-        self.tty = ""
-        self.id = 0
-        self.angle = 0
-        self.pos = 0
+    """servo class to hold servo parameters"""
+    id: int
+    tty: str
+    angle: int = 0
+    pos: int = 0
 
 
-# leg class to hold servo positions
+@dataclass
 class Leg:
-    def __init__(self, num):
-        self.num = num
-        self.servo = [Servo(), Servo(), Servo()]
+    """leg class to hold servo positions"""
+    id: int
+    servos: List[Servo]
+
+
+class ServoNotFoundError(EnvironmentError):
+    pass
+
+
+def get_servo_tty(servo_id, controllers):
+    for tty, controller in controllers.items():
+        try:
+            controller.get_servo_id(servo_id, timeout=0.05)
+            rospy.loginfo(f"found servo id {servo_id} on {tty}")
+            return tty
+        except TimeoutError:
+            pass
+
+    rospy.logerr(f"couldn't find a servo with ID: {servo_id} on any serial bus!")
+    raise ServoNotFoundError
+
+
+def set_tty_to_low_latency():
+    rospy.loginfo("setting ports to low latency mode")
+
+    for tty in TTY_LIST:
+        os.system(f"setserial {tty} low_latency")
+
+
+def spawn_reading_threads(servos, controllers):
+    rospy.loginfo("spawning reading threads")
+
+    thread_list = []
+    for tty, controller in controllers.items():
+        thread_list.append(ListenerThread([servo for servo in servos if servo.tty == tty], controller))
+        thread_list[-1].start()
+
+    return thread_list
 
 
 class ListenerThread(Thread):
-    def __init__(self, tty):
+    def __init__(self, servos: List[Servo], controller: lewansoul_lx16a.ServoController):
         Thread.__init__(self)
-        self.tty = tty
+        self.servos = servos
+        self.controller = controller
 
     def run(self):
         while not rospy.is_shutdown():
-            for leg_num in range(6):
-                for servo_num in range(3):
-                    id = leg_list[leg_num].servo[servo_num].id
-                    port = leg_list[leg_num].servo[servo_num].tty
-                    if port == self.tty:
```

Comment:
this tty check is not present in the new version, why?

Reply:
There is, I just do it once when the threads are created